core/trino-main/src/main/java/io/trino/sql/planner/LabelEvaluator.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner;
+
+import com.google.common.collect.ImmutableMap;
+import io.trino.Session;
+import io.trino.operator.DriverYieldSignal;
+import io.trino.operator.Work;
+import io.trino.operator.project.PageProjection;
+import io.trino.spi.Page;
+import io.trino.spi.block.Block;
+import io.trino.spi.function.WindowIndex;
+import io.trino.spi.type.Type;
+import io.trino.sql.tree.Label;
+
+import java.util.List;
+import java.util.Map;
+import java.util.OptionalInt;
+import java.util.function.Supplier;
+
+import static io.airlift.slice.Slices.utf8Slice;
+import static io.trino.operator.project.SelectedPositions.positionsRange;
+import static io.trino.spi.predicate.Utils.nativeValueToBlock;
+import static io.trino.spi.type.BigintType.BIGINT;
+import static io.trino.spi.type.BooleanType.BOOLEAN;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.sql.planner.PhysicalValuePointer.CLASSIFIER;
+import static io.trino.sql.planner.PhysicalValuePointer.MATCH_NUMBER;
+import static io.trino.sql.tree.Label.EMPTY;
+import static io.trino.sql.tree.Label.PARTITION_END;
+import static io.trino.sql.tree.Label.PARTITION_START;
+import static java.util.Objects.requireNonNull;
+
+public class LabelEvaluator
```

Comment:
Move all the "runtime" pattern classes into another package. They don't belong in the planner